pkg/provider/azure/keyvault/keyvault.go

```diff
@@ -0,0 +1,259 @@
+package keyvault
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/keyvault/keyvault"
+	kvauth "github.com/Azure/azure-sdk-for-go/services/keyvault/auth"
+	"golang.org/x/crypto/pkcs12"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1alpha1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1alpha1"
+	smmeta "github.com/external-secrets/external-secrets/apis/meta/v1"
+	"github.com/external-secrets/external-secrets/pkg/provider"
+	"github.com/external-secrets/external-secrets/pkg/provider/schema"
+)
+
+type Azure struct {
+	kube       client.Client
+	store      esv1alpha1.GenericStore
+	baseClient *keyvault.BaseClient
+	namespace  string
+	iAzure     IAzure
+}
+
+type IAzure interface {
+	getKeyVaultSecrets(ctx context.Context, vaultName string, version string, secretName string, withTags bool) (map[string][]byte, error)
+}
+
+func init() {
+	schema.Register(&Azure{}, &esv1alpha1.SecretStoreProvider{
+		AzureKV: &esv1alpha1.AzureKVProvider{},
+	})
+}
+
+func (a *Azure) New(ctx context.Context, store esv1alpha1.GenericStore, kube client.Client, namespace string) (provider.Provider, error) {
+	anAzure := &Azure{
+		kube:      kube,
+		store:     store,
+		namespace: namespace,
+	}
+	anAzure.iAzure = anAzure
+	azClient, err := anAzure.newAzureClient(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	anAzure.baseClient = azClient
+	return anAzure, nil
+}
+
+// implement store.Client.GetSecret Interface.
+// retrieve a secret with the secret name defined in ref.Property in a specific keyvault with the name ref.Name.
+func (a *Azure) GetSecret(ctx context.Context, ref esv1alpha1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	version := ""
+	var secretBundle []byte
+
+	if ref.Version != "" {
+		version = ref.Version
+	}
+	secretName := ref.Property
```

Comment:
I think the `ref.Key` should be used to lookup the secret name, not `ref.Property`.

Reply:
Hi @moolen ,
`ref.Key` will be holding the Keyvault name , while `ref.Property` will be holding the secret name .

this is inspired from the similar AWS implementation here : https://github.com/itscontained/secret-manager/blob/master/pkg/store/aws/aws.go#L94

Reply:
>`ref.Key`  will be holding the Keyvault name

The vaultName should be part of the store spec ([see comment below](https://github.com/external-secrets/external-secrets/pull/22/files#r571432395))



The `ref.Property` is `"used to select a specific property of the Provider value (if a map), if supported"`. I.e. once you have a secret map ref.Property is used to lookup a key in the map. 
Here it's passed down to `baseClient.GetSecret` as the name of the secret.

..i tried to clarify that, i hope it's not too confusing 🤔

Reply:
aha , well , now I got your point .

however I have some concerns of such proposal for different reasons :

1- this won't give the possibility of aggregating secret values from different keyvaults into a consolidated k8s secret 
2-  it will somehow break the loose coupling between the secretStore (holding the connectivity information) and the externalSecrets (holding the data to be replicated). Imagine an application being developped .

without this 1X1 coupling , an application developer (unaware of the infrastructure) will be defining the externalSecret focusing on from where to get the secrets values (keyvault name , secret name) , while some other infrastructure aware guy (unaware of how)will define the secret store by giving the tenantId , service principals  (and the azure access point as you mentioned in the comment below) . i don't want to mix the responsibilities.

3- another use case that will be broken if we define the  vault name in the secret store , is having different applications accessing the same Azure tenant but grabbing different secrets from the same or different KVs , it will be enough to define the secretstore once (giving the access information) and then each application is responsible of defining from where to get its data.

Reply:
`kind: ExternalSecret
metadata:
  name: azure-externalsecret-multi
  namespace: secret-manager-user
spec:
  storeRef:
    name: azure-secretstore
  data:
  - secretKey: appPassword
    remoteRef:
      name: kv-appsec-secretmg-mvp-passwords
      property: appPassword
  - secretKey: appuser
    remoteRef:
      name: kv-appsec-secretmg-mvp-users
      property: appuser
`

this is an example of the aggregation i meant in the first point of my comment above

Reply:
Regarding (1) and your example above: 
We want to do this with managed fields ([see docs](https://kubernetes.io/docs/reference/using-api/server-side-apply/#field-management)). This should enable it to savely sync two `Kind=ExternalSecret` into one `Kind=Secret`.

Regarding 2+3:
Conceptually, the `SecretStore` represents an instance of an k/v API that serves secrets (or: a part of infrastructure). and the `ExternalSecret` is the key that is looked up in that particular instance.
If i understand your use-case (2) correctly, your infrastructure guy would need to create multiple `Kind=SecretStore` (instead of just one) and the devs would have to reference those stores instead of using a `ref.Key` in an ExternalSecret to access. So the responsibilities aren't mixed up and (for 3.) multiple stores serve multiple users. 

Would that work for you? I think it's really important to clarify your use-case and - if necessary - let's draw/document them.

Reply:
Thanks for your reply. I understand you point for having the Azure KV name in the secret store . I ll try to return back to you to see if this can cover our use cases .

meanwhile , can you provide some examples for how the `ref.Key` and `ref.property` can be used in the Azure usecase ? because i was under the impression that the map returned is the list of ALL secrets saved in a specific keyvault (i referred to it by `ref.Key` , and the property is the specific secret we need to select from this map)

also i am wondering how the `getSecretMap` implementation will be working when using the `dataFrom` ? what will be the input passed in the `ref.Key` i was assuming that we will pass here the Azure Keyvault name as it is ' the key used in the Provider' 

Sorry for all my comments :-)  I think the confusion i am having is because we started this PR based on the itscontained/SecretManager Project and we we tried to follow an implementation similar to the [AWS implementation ](https://github.com/itscontained/secret-manager/blob/master/pkg/store/aws/aws.go#L110)which was using `ref.Name` as the secret manager name.

I will be more than happy to draw/document use cases and as mentioned , examples for how the CRs will look like should be  the best way to put all of them on the table.

Reply:
Your vault `kv-appsec-secretmg-mvp` is (simplified) a key-value store: `appPassword`, `dbuser` etc. are lookup-keys and hold simple values and metadata. The `ref.Key` should match the lookup-keys. 

People don't always store simple text in a `value`. Some APIs (or users) use JSON. So if `ref.Property` is defined, the `value` is parsed as JSON (we expect a json map) and `ref.Property` is used to lookup a key in this particular json map. 
That's how it's implemented in KES: https://github.com/external-secrets/kubernetes-external-secrets/blob/master/lib/backends/kv-backend.js#L50-L62

If you've worked with AWS SecretsManager before, it's actually part of the UI and [API](https://docs.aws.amazon.com/sdk-for-go/api/service/secretsmanager/#GetSecretValueOutput).

So your example could look like this:

Vault `kv-appsec-secretmg-mvp`
lookup-key | value |
--- | --- |
appPassword | `1234` | 
myJSON | `{"foo": "bar", "baz": "bang"}` | 

You could use `ref.Key=myJSON` and `ref.Property=baz` to extract `bang` from the vault.

Reply:
Let me try to jump into this @moolen . The API structure in Azure is a little bit different from the one of AWS. Here are the references we used to figure-out the differences:

* https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_CreateSecret.html
* https://docs.microsoft.com/en-us/rest/api/keyvault/

From what I understood from the AWS API, the secret-store is unique per secret & contains several secrets only because it API is used to encode multiple secrets at once using JSON. That make the approach of taking every JSON property as secret names & their associated values at once perfectly valid. The mapping to K8's secret structure as secretName1: secretValue1, secretName2: secretValue2, secretName3: secretValue3 then also comes as perfectly natural (it is even documented as such in AWS's API doc). We then assume the SSM secret name is the Secret-store name & the secret's JSON property as the secret keys.

In the Azure API, there is an additional layer that we want to leverage: each secret is a separated Azure object in the Vault (which is an object by itself), with its own API & versioning. Each secret can be accessed separately, but the Vault also offers the possibility to grab every secrets at once, which provide a similar convenience as AWS's JSON encoding.

Given that Azure data-model has fine-grained access to individual secrets, it would not make sense to bundle every secrets in one as we do in AWS. Still, bundling makes sense for AWS. I tend to believe we can have both model with the same API: what matters is that this Application finally gets its secretName1: secretValue1, secretName2: secretValue2, secretName3: secretValue3.

To achieve this, here is a proposal:

1. SecretStore name is the SSM secretName in AWS & the KeyVault name in Azure
2. Individual Secret strings are stored in the ExternalSecret, but they refer to different storage in AWS (JSON keys in the SSM) and in Azure (Secret object in the KeyVault Object).
3.  Obviously, the above implies that secrets bundling+encoding then becomes a provider-specific code

WDYT?

Reply:
I think to stick with the API conventions of the ExternalSecret, the optional `property` value should remain as a way to obtain nested values (json or otherwise) inside the referenced provider secret.

In this case for Azure, we could follow the "directory-style" (similar to Hashicorp Vault) method of obtaining the secret within the Azure Vault, where `remoteRef.name`'s end up being paths to the secret within the system. So @ahmedmus21's example above would become -

```yaml
kind: ExternalSecret
metadata:
  name: azure-externalsecret-multi
  namespace: secret-manager-user
spec:
  storeRef:
    name: azure-secretstore
  data:
  - secretKey: appPassword
    remoteRef:
      name: kv-appsec-secretmg-mvp-passwords/appPassword
      # property is still available to obtain nested values within the fetched secret
  - secretKey: appuser
    remoteRef:
      name: kv-appsec-secretmg-mvp-users/appuser
```

Reply:
As it is today in the PR, the storeRef's is not a store, but a set of (ref-to) credentials to access the store. The Vault itself it really `kv-appsec-secretmg-mvp-passwords`, which is not a folder in the Vault. 

In addition the Azure SDK requires a full URL to the vault (domain may varry today & will even more when Azure will enable custom domain for KeyVault). So, If we use the Azure Vault name directly in ExternalSecret, we have a value for the Azure KeyVault name both in the ExternalSecret & in the SecretStore (the URL)... so room for inconsistency.

Can you point me to _the API conventions of the ExternalSecret_ ?

Reply:
> Can you point me to the API conventions of the ExternalSecret ?

The specification of the ExternalSecret was moved to a [separate repo](https://github.com/external-secrets/crd-spec/blob/main/Spec.md#external-secret) after many threads in the original kubernetes-external-secrets project. We also do our best effort to follow [Kubernetes API conventions](https://github.com/kubernetes/community/blob/6254e11696ac0886d554021a27c8179b459502c6/contributors/devel/sig-architecture/api-conventions.md#api-conventions).

The idea behind the `SecretStore` resource is to separate concerns authentication/access and the actual Secret and configuration needed for workloads. The `ExternalSecret` specifies what to fetch, the `SecretStore` specifies how to access.

I think its fine that the `kv-appsec-secretmg-mvp-passwords` is not actually a directory, I was just referencing the path based directory style, as other providers are more directory based and will also share that style. The Azure VaultEndpoint can be formed by combining the `vaultEndpoint` in the SecretStore, with the first section of the `remoteRef.name`.  This also has the benefit that if `kv-appsec-secretmg-mvp-passwords/appPassword` is a secret in json form, then the `property` field can be used to extract field(s) from the fetched value.

Reply:
Actually, an Azure KeyVault instance is much like an HashiCorp Vault instance, with the difference  that it does not offer folders to support complex RBAC. Several Vault instances may sit in several regions, with each having a separated set of users & content. So there is not such thing a central "Store" encompassing every Key-Vaults, under which the Vault names would be just folders.

![image](https://user-images.githubusercontent.com/73526813/108370755-48625300-71fd-11eb-833c-7f04a6f3e43b.png)

For the reasons above, we believe that a Vault instance is a kind of `SecretStore`, and the above-mentioned `kv-appsec-secretmg-mvp-passwords` should be the name of a `SecretStore` of kind Azure KeyVault, rather than a folder in the `ExternalSecret`.

Anything we are missing?

Reply:
That works for me, I am ok with either `vaultName` being a field in the azure vault `SecretStore`, or being the first path of the `ExternalSecret` key. Reading through the current solution in this PR has the `vaultName` being set by `remoteRef.key` and the secret name being set by `remoteRef.property`.

`vaultName` (or equivalent) as a field in `SecretStore` seems to make sense, as that is pretty similar to Hashicorp vault namespace that can be optionally set on the Hashicorp (enterprise) vault SecretStore fields.

I believe our main concern is the use of the `remoteRef.property` field, which should be limited to optional extraction of data after the secret is fetched, but not used as part of the secret fetching process.

Reply:
wrt/ vaultName: Indeed, we are preparing a PR update to fix this & also a (prior) PR the CRD spec to show it before we align the code on it.

wrt/ remoteRef Ok, I get your point. As you can see in the Object model above, Secret are individually-managed objects (with their own version & all). However, Azure offers the possibility to fetch them all at once, which is highly desirable as (1) an application often consumes every secrets at once (at Pod startup) and (2) Azure billing is per-API call. So having 3x fetch calls for 3 secret objects causes (1) avoidable delay as all of them need to fetched before the stuff starts and (2) unnecessary billing. So why not making the secret extraction a provider-specific code? Obviously the AWS provider would use JSON bundling, while Azure's would be using separated objects (which have a type on their own). Another secret vault (CyberARK) also come with individual strongly typed secrets (they are called "Accounts") that can be fetched all-at-once from their store (called "Safe")..

Reply:
I moved the CRD-level discussion for the Vault Name to this PR: https://github.com/external-secrets/crd-spec/pull/5